### PR TITLE
Fixes for authentication plugin events not fired for non-existent users

### PR DESCRIPTION
**Pull request by jon5477**
Fixes #27836
Fixes #29517

### CODE REVIEW

File: login_password_page.php [hunk 1]
```diff
@@ -82,24 +82,29 @@
 # Get the user id and based on the user decide whether to continue with native password credential
 # page or one provided by a plugin.
 $t_user_id = auth_get_user_id_from_login_name( $t_username );
-if( $t_user_id !== false && auth_credential_page( '', $t_user_id ) != AUTH_PAGE_CREDENTIAL ) {
+# User id could be false if the user does not exist in DB, should be calling auth_credential_page
+# regardless if the user exists or not to give the plugin an opportunity to handle non-existent
+# users per #29517
+$t_should_redirect = ($t_user_id !== false ? auth_credential_page('', $t_user_id) : auth_credential_page('', NO_USER, $t_username)) != AUTH_PAGE_CREDENTIAL;
```
Comment (dregad) [inline login_password_page.php:88]: Break long line

File: login_password_page.php [hunk 2]
```diff
@@ -82,24 +82,29 @@
...
+	if( !is_blank( $f_cookie_error ) ) {
 		$t_query_args['cookie_error'] = $f_cookie_error;
-    }
+	}
 
 	$t_query_text = http_build_query( $t_query_args, '', '&' );
 
-	$t_redirect_url = auth_credential_page( $t_query_text, $t_user_id );
+	# Determine the credential page URL based on user id (if it exists) or username
+	$t_redirect_url = $t_user_id !== false ? auth_credential_page($t_query_text, $t_user_id) : auth_credential_page($t_query_text, NO_USER, $t_username);
```
Comment (dregad) [inline login_password_page.php:107]: Break long line

**Review by dregad — CHANGES_REQUESTED** [review]
Thanks for your contribution @jon5477.

Please make sure that your submissions follows our [coding guidelines](https://mantisbt.org/wiki/doku.php/mantisbt:coding_guidelines).

Also, since you mention this fixes #27836, kindly review PR #1712, which was submitted as an attempt to address the same issue, and confirm whether your work replaces it, and as mentioned in [my note in #29517](https://mantisbt.org/bugs/view.php?id=29517#c66205) maybe the issue itself is a duplicate.

**Comment by jon5477** [comment]
I can confirm that this PR is a fix for the same issue in #1712, but this maintains code compatibility of an older fix made in [#25061](https://mantisbt.org/bugs/view.php?id=25061). This PR also fixes an issue that a user made a [comment](https://github.com/mantisbt/mantisbt/pull/1712#issuecomment-800305017) upon in #1712.

File: core/authentication_api.php [hunk 3]
```diff
@@ -81,14 +81,16 @@
...
  *                                  in the system, that may be auto-provisioned.
- * @param string        $p_username The username or email
+ * @param string|null   $p_username The username or email
  * @return AuthFlags The auth flags object to use.
  * @throws ClientException
  */
 function auth_flags( $p_user_id = null, $p_username = '' ) {
-	if( !$p_user_id ) {
+	# If user id is null but username is set, let the plugin handle it per #27836
+	if (is_null($p_user_id) || ($p_user_id === false && is_null($p_username))) {
```
Comment (dregad) [inline core/authentication_api.php:90]: Coding guidelines: you missed that one...

File: login_password_page.php [hunk 4]
```diff
@@ -82,24 +82,33 @@
 # Get the user id and based on the user decide whether to continue with native password credential
 # page or one provided by a plugin.
 $t_user_id = auth_get_user_id_from_login_name( $t_username );
-if( $t_user_id !== false && auth_credential_page( '', $t_user_id ) != AUTH_PAGE_CREDENTIAL ) {
+# User id could be false if the user does not exist in DB, should be calling auth_credential_page
+# regardless if the user exists or not to give the plugin an opportunity to handle non-existent
+# users per #29517
+$t_should_redirect = ($t_user_id !== false ?
+		auth_credential_page( '', $t_user_id ) :
+		auth_credential_page( '', NO_USER, $t_username )) != AUTH_PAGE_CREDENTIAL;
```
Comment (dregad) [inline login_password_page.php:90]:
```php
$t_should_redirect = AUTH_PAGE_CREDENTIAL != ( $t_user_id !== false 
		? auth_credential_page( '', $t_user_id )
		: auth_credential_page( '', NO_USER, $t_username ) );
```

File: login_password_page.php [hunk 5]
```diff
@@ -82,24 +82,33 @@
...
-    }
+	}
 
 	$t_query_text = http_build_query( $t_query_args, '', '&' );
 
-	$t_redirect_url = auth_credential_page( $t_query_text, $t_user_id );
+	# Determine the credential page URL based on user id (if it exists) or username
+	$t_redirect_url = $t_user_id !== false ?
+		auth_credential_page( $t_query_text, $t_user_id ) :
+		auth_credential_page( $t_query_text, NO_USER, $t_username );
```
Comment (dregad) [inline login_password_page.php:111]: Put `?` and `:` at begining of line, as per review comment for line 90 above

**Review by dregad — CHANGES_REQUESTED** [review]
A few more guidelines issues to fix.

Note: I can't actually test this properly as I don't have an environments with auth plugins, and no time to set one up, so I'll take your word for it assuming you've done your homework and made sure your code works and does not introduce regressions (which does not seem to be the case based on my review, but I may have missed something).

**Review by dregad — APPROVED** [review]
Thanks. Waiting a few days to give others a chance to review before I merge.

**Pull request merged**